src/Tutorials/Custom-Grain-Storage.md

```diff
@@ -0,0 +1,232 @@
+{
+    var fName = grainReference.ToKeyString() + "." + grainType;
+    var path = Path.Combine(_rootDirectory, fName);
+
+    var fileInfo = new FileInfo(path);
+    if (!fileInfo.Exists)
+        return;
+
+    using (var stream = fileInfo.OpenText())
+    {
+        var storedData = await stream.ReadToEndAsync();
+        grainState.State = JsonConvert.DeserializeObject(storedData, _jsonSettings);
+    }
+}
+```
+
+Note that for the deserialization, we use the `_jsonSettings` which was set on the `Init` function. This is important to be able to serialize/deserialize properly the state.
+
+## Writing State
+
+Writing the state is similar to reading the state.
+
+```csharp
+public async Task WriteStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+{
+    var storedData = JsonConvert.SerializeObject(grainState.State, _jsonSettings);
+
+    var fName = grainReference.ToKeyString() + "." + grainType;
+    var path = Path.Combine(_rootDirectory, fName);
+
+    var fileInfo = new FileInfo(path);
+
+    using (var stream = new StreamWriter(fileInfo.Open(FileMode.Create, FileAccess.Write)))
+    {
+        await stream.WriteAsync(storedData);
+    }
+}
+```
+
+Similarly as reading, we use `_jsonSettings` to write the state.
+
+## Clearing State
+
+Clearing the state would be deleting the file if the file exists.
+
+```csharp
+public Task ClearStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+{
+    var fName = grainReference.ToKeyString() + "." + grainType;
+    var path = Path.Combine(_rootDirectory, fName);
+
+    var fileInfo = new FileInfo(path);
+    if (fileInfo.Exists)
+        fileInfo.Delete();
+
+    return Task.CompletedTask;
+}
+```
+
+## Putting it Together
+
+In order to use the storage, we should be able to configure the root directory. For that we will create an options file `FileGrainStorageOptions`:
+
+```csharp
+public class FileGrainStorageOptions
+{
+    public string RootDirectory { get; set; }
+}
+```
+
+After that we will create a factory which will allow us to scope the options setting to the provider name and at the same time create an instance of the `FileGrainStorage` to ease the registration to the service collection.
+
+```csharp
+public static class FileGrainStorageFactory
+{
+    internal static IGrainStorage Create(IServiceProvider services, string name)
+    {
+        IOptionsSnapshot<FileGrainStorageOptions> optionsSnapshot = services.GetRequiredService<IOptionsSnapshot<FileGrainStorageOptions>>();
+        var options = optionsSnapshot.Get(name);
+        return ActivatorUtilities.CreateInstance<FileGrainStorage>(services, name, options.RootDirectory);
+    }
+}
+```
+
+Lastly to register the grain storage, we create an extension on the `ISiloHostBuilder` which internally register the grain storage as a named service using `.AddSingletonNamedService(...)`, an extension provided by `Orleans.Core`.
```

Comment:
May want to be clear that two named services need be added, one for the named IGrainStorage service, and one for the named ILifecycleParticipant<ISiloLifecycle> service.

Reply:
Thanks! I added an extra explanation on this.